test/validation/api/pktio/pktio.c

```diff
@@ -2503,6 +2503,245 @@ static void pktio_test_pktout_ts(void)
+
+	/* Receive Packet Tx completion events for all sent/dropped pkts */
+	for (i = 0; i < TX_BATCH_LEN; i++) {
+		if (use_plain_queue) {
+			ev = odp_queue_deq(compl_queue[i]);
+
+			/* Event validation */
+			CU_ASSERT_FATAL(ev != ODP_EVENT_INVALID);
+			CU_ASSERT_FATAL(odp_event_type(ev) == ODP_EVENT_PACKET_TX_COMPL);
+			CU_ASSERT_FATAL(odp_packet_tx_compl_from_event(ev) !=
+					ODP_PACKET_TX_COMPL_INVALID);
+
+			tx_compl = odp_packet_tx_compl_from_event(ev);
+			CU_ASSERT_FATAL(odp_packet_tx_compl_to_event(tx_compl) == ev);
+
+			/* User ptr should be same as packet's user ptr */
+			CU_ASSERT(odp_packet_tx_compl_user_ptr(tx_compl) ==
+				  (const void *)&pkt_seq[i]);
+
+			/* Alternatively call event free / compl free */
+			if (i % 2)
+				odp_packet_tx_compl_free(tx_compl);
+			else
+				odp_event_free(ev);
+		} else {
+			odp_queue_t rcv_queue;
+			int j;
+
+			ev = odp_schedule(&rcv_queue, wait);
+
+			/* Event validation */
+			CU_ASSERT_FATAL(ev != ODP_EVENT_INVALID);
+			CU_ASSERT_FATAL(odp_event_type(ev) == ODP_EVENT_PACKET_TX_COMPL);
+			CU_ASSERT_FATAL(odp_packet_tx_compl_from_event(ev) !=
+					ODP_PACKET_TX_COMPL_INVALID);
+
+			tx_compl = odp_packet_tx_compl_from_event(ev);
+			CU_ASSERT_FATAL(odp_packet_tx_compl_to_event(tx_compl) == ev);
+
+			/* User ptr should be same as packet's user ptr i.e seq array ptr */
+			for (j = 0; j < TX_BATCH_LEN; j++) {
+				if (!seq_found[j] &&
+				    ((const void *)&pkt_seq[j] ==
+				     odp_packet_tx_compl_user_ptr(tx_compl))) {
+					/* Mark that sequence number is found */
+					seq_found[j] = 1;
+
+					/* Receive queue validation */
+					CU_ASSERT(rcv_queue == compl_queue[j]);
+					break;
+				}
+			}
+			/* Check that sequence number is found */
+			CU_ASSERT(j < TX_BATCH_LEN);
+
+			/* Alternatively call event free / compl free */
+			if (i % 2)
+				odp_packet_tx_compl_free(tx_compl);
+			else
+				odp_event_free(ev);
+		}
+	}
+
+	for (i = 0; i < TX_BATCH_LEN; i++)
+		odp_queue_destroy(compl_queue[i]);
+
+	for (i = 0; i < num_ifaces; i++) {
+		CU_ASSERT_FATAL(odp_pktio_stop(pktio[i]) == 0);
+		CU_ASSERT_FATAL(odp_pktio_close(pktio[i]) == 0);
+	}
+}
+
+static int pktio_check_pktout_compl_plain_queue(void)
+{
+	odp_pktio_param_t pktio_param;
+	odp_pktio_capability_t capa;
+	odp_pktio_t pktio;
+	int idx = 0;
+	int ret;
+
+	odp_pktio_param_init(&pktio_param);
+	pktio_param.in_mode = ODP_PKTIN_MODE_DIRECT;
+	pktio_param.out_mode = ODP_PKTOUT_MODE_DIRECT;
+
+	pktio = odp_pktio_open(iface_name[idx], pool[idx], &pktio_param);
+	if (pktio == ODP_PKTIO_INVALID)
+		return ODP_TEST_INACTIVE;
+
+	ret = odp_pktio_capability(pktio, &capa);
+	(void)odp_pktio_close(pktio);
+
+	if (ret < 0 || !capa.tx_compl.mode_all || !capa.tx_compl.queue_type_plain)
+		return ODP_TEST_INACTIVE;
+
+	return ODP_TEST_ACTIVE;
+}
+
+static int pktio_check_pktout_compl_sched_queue(void)
+{
+	odp_pktio_param_t pktio_param;
+	odp_pktio_capability_t capa;
+	odp_pktio_t pktio;
+	int idx = 0;
```

Comment:
Same here. Actually, everything else is the same in this function as in plain queue check, but comparison with capa.tx_compl.queue_type_sched vs. capa.tx_compl.queue_type_plain. You could reuse check function code the same way as the test function is reused.

```
static int pktio_check_pktout_compl(bool plain)
{
 ...
}

static int pktio_check_pktout_compl_sched(void)
{
     return pktio_check_pktout_compl(false);
}

static int pktio_check_pktout_compl_plain(void)
{
     return pktio_check_pktout_compl(true);
}

```